### PR TITLE
expose get url method for wrapped signal

### LINKED ISSUE #22: expose method to get url from signalName + parameters
We should be able to make traditional links based on routing config since it allows to use `rightClick + copy` or `rightClick + open in new tab` actions.

It could be achieved with method to get url from signalName and parameters. So it would be able to create something like `Link` component @jeroenverfallie mentioned in #14 and without extra router configuration
```jsx
<Link signal="someSignal" args={ { someParam: true } } >
```
 which will render to 
```jsx
<a href="/#/url/param" onClick=(e) = { e.preventDefault(); this.props.signals.someSignal( { someParam: true })}>
```

### LINKED ISSUE #36: payload.url seems useless in current setup
https://github.com/christianalfoni/cerebral-router/blob/8daa9987ac5177ebaa3c3f6f3bfd64a466d2a699/index.js#L52

* `payload.url` always undefined when triggered from url
* `payload.url` could be set within signal call but its value never used

### LINKED ISSUE #37: router could break signal chain
Let assume developer used `url`, `path`, `params` or `query` key in input. [Chain inputs would be broken](https://github.com/christianalfoni/cerebral-router/blob/8daa9987ac5177ebaa3c3f6f3bfd64a466d2a699/index.js#L51-L60) once router added to project.
Also this keys can't be used as keys for route definitions.

### LINKED ISSUE #39: redundant strip off baseUrl
https://github.com/christianalfoni/cerebral-router/blob/285c9c36dd3e73507e137ffa06836918246e0b20/index.js#L103
seems redundant since url in this place newer contains baseUrl. Could break app if will accidentally match with part of constructed url.

### LINKED ISSUE #43: es6 specific syntax
Function.name is part of ES6 with poor browser support

https://github.com/christianalfoni/cerebral-router/commit/8384260bab394aa73de70a2332e557013eb4e684#diff-168726dbe96b3ce427e7fedce31bb0bcR28

### PR DESCRIPTION
**Pull request by Guria**
resolves #22 
resolves #39 
resolves #43 

includes #41 which resolves #37 and resolves #36

does nothing with #34
need detailed review before merge

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -36,79 +83,39 @@ function router (controller, routes, options) {
       throw new Error('Cerebral router - The signal "' + routes[route] + '" for the route "' + route + '" does not exist.');
     }
 
-    if (signal.name === 'wrappedSignal') {
+    if (typeof signal.getUrl === "function") {
```
Comment (Guria) [inline index.js]: #43

File: index.js [hunk 2]
```diff
@@ -36,79 +83,39 @@ function router (controller, routes, options) {
...
-        }
-      }
+    // callback for urlMapper
+    wrappedRoutes[route] = function(payload) {
+      wrappedSignal({ route: payload });
+    };
 
-      url = url === '*' ? location.pathname : url;
-      url = options.onlyHash && url.indexOf('#') === -1 ? '/#' + url : url;
-      input.route.url = options.baseUrl && url.substr(0, options.baseUrl.length) === options.baseUrl ? url.replace(options.baseUrl, '') : url;
```
Comment (Guria) [inline index.js]: #39

File: index.js [hunk 3]
```diff
@@ -36,79 +83,39 @@ function router (controller, routes, options) {
...
-      var payload = hasSync ? arguments[1] : arguments[0] || {};
-
-      var input = payload;
-      input.route = {
-        url: payload.url,
-        path: payload.path,
-        params: payload.params,
-        query: payload.query
-      };
-      delete input.url;
```
Comment (Guria) [inline index.js]: #37

File: index.js [hunk 4]
```diff
@@ -36,79 +83,39 @@ function router (controller, routes, options) {
...
+    }
 
-        if (!urlMatched) {
-          throw new Error('Cerebral router - Computed url for signal "' + routes[route] +'" can\'t match given route "' + route + '".\n' +
-                          'Check required params provided to signal is not falsy.');
-        }
-      }
+    // callback for urlMapper
+    wrappedRoutes[route] = function(payload) {
+      wrappedSignal({ route: payload });
```
Comment (Guria) [inline index.js]: #37

**Comment by christianalfoni** [comment]
Okay,

I have gone through the changes and to me they seem very reasonable. It is also a cleaner implementation :-) 

I must admit I am having a hard time composing all related issues to the changes in code and what they might affect... but that is why we have tests. So I will merge and we use this as a base for securing scenarios with more tests before release.

Thanks for your efforts on this @Guria and sorry it takes so much time, but I simply am in lack of time these days :-/

**Pull request merged**